### PR TITLE
Replace NotNullAttribute with thrown exceptions

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/ValidationSummaryTagHelperTest.cs [hunk 1]
```diff
@@ -279,10 +279,10 @@ public class ValidationSummaryTagHelperTest
             var generator = new TestableHtmlGenerator(new EmptyModelMetadataProvider());
 
             var validationSummaryTagHelper = new ValidationSummaryTagHelper(generator);
-            var expectedMessage = string.Format(
-                @"The value of argument 'value' ({0}) is invalid for Enum type 'Microsoft.AspNet.Mvc.ValidationSummary'.
-Parameter name: value",
-                validationSummary);
+            var validationTypeName = typeof(ValidationSummary).FullName;
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/ValidationSummaryTagHelperTest.cs]: cc @rynowak

**Comment by pranavkm** [comment]
I'll check in the changes to unbreak the CI but keep the PR open.

File: src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorParser.cs [hunk 2]
```diff
@@ -33,12 +32,32 @@ public class MvcRazorParser : RazorParser
...
+            RazorParser parser,
+            IReadOnlyList<ChunkTree> inheritedChunkTrees,
+            IReadOnlyList<Chunk> defaultInheritedChunks,
+            string modelExpressionTypeName)
             : base(parser)
         {
+            if (parser == null)
+            {
+                throw new ArgumentNullException(nameof(parser));
+            }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorParser.cs]: Doesn't `RazorParser`'s constructor check `parser`?

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs [hunk 3]
```diff
@@ -41,10 +45,20 @@ public CompilerCache([NotNull] IFileProvider fileProvider)
...
-            [NotNull] IFileProvider fileProvider,
-            [NotNull] IDictionary<string, Type> precompiledViews)
+            IFileProvider fileProvider,
+            IDictionary<string, Type> precompiledViews)
             : this(fileProvider)
         {
+            if (fileProvider == null)
+            {
+                throw new ArgumentNullException(nameof(fileProvider));
+            }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs]: Checked in prior constructor

File: src/Microsoft.AspNet.Mvc.Razor/DefaultViewLocationCache.cs [hunk 4]
```diff
@@ -40,9 +45,14 @@ public ViewLocationCacheResult Get([NotNull] ViewLocationExpanderContext context
 
         /// <inheritdoc />
         public void Set(
-            [NotNull] ViewLocationExpanderContext context,
-            [NotNull] ViewLocationCacheResult value)
+            ViewLocationExpanderContext context,
+            ViewLocationCacheResult value)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/DefaultViewLocationCache.cs]: Nice. Does your module detect value types and ignore `[NotNull]` when removing it?

File: src/Microsoft.AspNet.Mvc.Razor/RazorView.cs [hunk 5]
```diff
@@ -75,8 +74,18 @@ private bool EnableInstrumentation
...
         {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/RazorView.cs]: Huh? Why is this block repeated?

File: src/Microsoft.AspNet.Mvc.Razor/ViewLocationCacheResult.cs [hunk 6]
```diff
@@ -21,10 +21,20 @@ public struct ViewLocationCacheResult : IEquatable<ViewLocationCacheResult>
...
         {
+            if (foundLocation == null)
+            {
+                throw new ArgumentNullException(nameof(foundLocation));
+            }
+
+            if (searchedLocations == null)
+            {
+                throw new ArgumentNullException(nameof(searchedLocations));
+            }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/ViewLocationCacheResult.cs]: Checked in following constructor i.e. before we reach here.

File: test/Microsoft.AspNet.Mvc.Razor.Test/RazorPageTest.cs [hunk 7]
```diff
@@ -1060,7 +1059,7 @@ public void WriteAttributeTo_WritesAsExpected(AttributeValue[] attributeValues,
                     },
                     {
                         GetTagHelperOutput(
-                            tagName:     null,
+                            tagName:     "",
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Razor.Test/RazorPageTest.cs]: `string.Empty` :smiling_imp:

File: src/Microsoft.AspNet.Mvc.Razor/ViewLocationExpanderContext.cs [hunk 8]
```diff
@@ -18,10 +17,20 @@ public class ViewLocationExpanderContext
         /// <param name="actionContext">The <see cref="Mvc.ActionContext"/> for the current executing action.</param>
         /// <param name="viewName">The view name.</param>
         /// <param name="isPartial">Determines if the view being discovered is a partial.</param>
-        public ViewLocationExpanderContext([NotNull] ActionContext actionContext,
-                                           [NotNull] string viewName,
+        public ViewLocationExpanderContext(ActionContext actionContext,
+                                           string viewName,
                                            bool isPartial)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/ViewLocationExpanderContext.cs]: Think this is the only fancy indentation visible in this PR. Great to see all the other cleanup.

**Comment by dougbu** [comment]
AppVeyor says tests still hit `ArgumentNullException`s.

**Comment by dougbu** [comment]
Realize this is already in so my :shipit: is meaningless. My comments are mostly about minor redundancies. The CI failures and the duplicate check in `RazorView` are important however.

File: src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorParser.cs (hunk 2 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorParser.cs]: no idea.

File: src/Microsoft.AspNet.Mvc.Razor/DefaultViewLocationCache.cs (hunk 4 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/DefaultViewLocationCache.cs]: Yup.

File: src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorParser.cs (hunk 2 above)
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorParser.cs]: I could see this either way.

**Pull request closed without merging**